### PR TITLE
feat: use temp file for snapshot [fetch|export] before finished

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- use temp file for `snapshot [fetch|export]` before finished
- extend `snapshot [list|remove|prune|clean|]` logic to handle those temp files

**Reference issue to close (if applicable)**

Closes https://github.com/ChainSafe/forest/issues/2101

**Other information and links**

### CODE REVIEW

**Review by lemmih — APPROVED** [review]
Looks good. I wonder if we can resume a partial download that got interrupted.

**Review by LesnyRumcajs — COMMENTED** [review]
When creating the issue I was instead thinking about using a *more* temporary file, i.e. writing it somewhere to `/tmp`. This way, in case of an interrupt or a panic it would get automatically cleaned up by the OS on the next reboot.

**Comment by hanabi1224** [comment]
> i.e. writing it somewhere to /tmp

There are some corner cases like when the temp folder and snapshot folder are on different drive, the `rename` can be expensive

> get automatically cleaned up by the OS

Not sure if this works on all OS, e.g. windows?

File: forest/cli/src/cli/snapshot_cmd.rs [hunk 1]
```diff
@@ -316,13 +322,20 @@ fn clean(config: &Config, snapshot_dir: &Option<PathBuf>, force: bool) -> anyhow
     let snapshots_to_delete: Vec<_> = fs::read_dir(snapshot_dir)?
         .flatten()
         .map(|entry| entry.path())
-        .filter(|p| p.extension().unwrap_or_default() == "car")
+        .filter(|p| {
+            let ext = p.extension().unwrap_or_default();
+            ext == "car" || ext == "tmp"
```
Comment (LesnyRumcajs) [inline forest/cli/src/cli/snapshot_cmd.rs:327]: How about moving this predicate to a common method?

**Review by tyshko5 — APPROVED** [review]

**Review by LesnyRumcajs — COMMENTED** [review]
To my understanding, this will still bloat the filesystem in case of a failure that is not even a panic. E.g. if something makes the method return an `Err` just before the rename. While I understand some artefacts may be left on a `panic` somewhere in the middle, they shouldn't be there on a gracious failure.

If you don't feel like using `TempDir` or `TempFile` are a good idea, perhaps it'd be worth writing a custom wrapper with a deleter in the destructor

For the `snapshot fetch`, it doesn't seem to me that the user is sufficiently informed about the artefact, somewhere on the disk.

File: forest/cli/src/cli/snapshot_cmd.rs (hunk 1 above)
Comment (hanabi1224) [reply forest/cli/src/cli/snapshot_cmd.rs]: Fixed

**Comment by hanabi1224** [comment]
> perhaps it'd be worth writing a custom wrapper with a deleter in the destructor

@lemmih that's great idea, fixed

File: forest/cli/src/cli/snapshot_cmd.rs [hunk 2]
```diff
@@ -217,7 +222,8 @@ fn remove(config: &Config, filename: &PathBuf, snapshot_dir: &Option<PathBuf>, f
     let snapshot_path = snapshot_dir.join(filename);
     if snapshot_path.exists()
         && snapshot_path.is_file()
-        && snapshot_path.extension() == Some(OsStr::new("car"))
+        && (snapshot_path.extension() == Some(OsStr::new("car"))
+            || snapshot_path.extension() == Some(OsStr::new("tmp")))
```
Comment (lemmih) [inline forest/cli/src/cli/snapshot_cmd.rs:226]: Can you use `is_car_or_tmp` here?

Comment (hanabi1224) [reply forest/cli/src/cli/snapshot_cmd.rs]: Good catch, fixed

**Pull request merged**